llvm/lib/Transforms/InstCombine/InstCombineCalls.cpp

```diff
@@ -1601,6 +1601,15 @@ Instruction *InstCombinerImpl::visitCallInst(CallInst &CI) {
       return CastInst::Create(Instruction::ZExt, NarrowAbs, II->getType());
     }
 
+    // abs(a * abs(b)) --> abs(a * b)
+    Value *A, *B;
+    if (match(IIOperand,
+              m_OneUse(m_c_Mul(m_Value(A),
+                               m_Intrinsic<Intrinsic::abs>(m_Value(B)))))) {
+      Value *New = Builder.CreateMul(A, B);
```

Comment:
You can propagate `nsw` to the new mul: https://alive2.llvm.org/ce/z/WjjNFF

Reply:
This is still outstanding.

Reply:
@goldsteinn, but it doesn't work here: https://alive2.llvm.org/ce/z/kXQA3p

Reply:
Ah, seems to only be if the abs precludes INT_MIN. (outer abs)

Reply:
can you add this?

Reply:
@goldsteinn, updated